File: src/dojopool/frontend/types/venue.ts
export enum VenueStatus {
    ACTIVE = 'ACTIVE',
    INACTIVE = 'INACTIVE',
    MAINTENANCE = 'MAINTENANCE',
    CLOSED = 'CLOSED'
}

export enum TableType {
    STANDARD = 'STANDARD',
    TOURNAMENT = 'TOURNAMENT',
    SNOOKER = 'SNOOKER',
    PRACTICE = 'PRACTICE'
}

export enum TableStatus {
    AVAILABLE = 'AVAILABLE',
    OCCUPIED = 'OCCUPIED',
    RESERVED = 'RESERVED',
    MAINTENANCE = 'MAINTENANCE'
}

export interface PoolTable {
    id: string;
    number: number;
    type: TableType;
    status: TableStatus;
    size: string;
    condition: number; // 1-5 rating
    lastMaintenance?: Date;
    currentGame?: {
        id: string;
        startTime: Date;
        players: string[]; // player IDs
    };
    nextReservation?: {
        id: string;
        startTime: Date;
        endTime: Date;
        players: string[]; // player IDs
    };
}

export interface BusinessHours {
    day: number; // 0-6, where 0 is Sunday
    open: string; // HH:mm format
    close: string; // HH:mm format
    closed: boolean;
}

export interface VenueAmenity {
    id: string;
    name: string;
    description?: string;
    icon?: string;
}

export interface VenuePricing {
    type: 'PER_HOUR' | 'PER_GAME' | 'MEMBERSHIP';
    amount: number;
    currency: string;
    description: string;
    peakHours?: {
        start: string; // HH:mm format
        end: string; // HH:mm format
        multiplier: number;
    };
}

export interface VenueStats {
    totalTables: number;
    availableTables: number;
    averageWaitTime: number; // in minutes
    currentOccupancy: number; // percentage
    popularHours: {
        hour: number;
        occupancy: number;
    }[];
}

export interface Venue {
    id: string;
    name: string;
    description: string;
    address: {
        street: string;
        city: string;
        state: string;
        country: string;
        postalCode: string;
        coordinates?: {
            latitude: number;
            longitude: number;
        };
    };
    contact: {
        phone: string;
        email: string;
        website?: string;
    };
    status: VenueStatus;
    businessHours: BusinessHours[];
    tables: PoolTable[];
    amenities: VenueAmenity[];
    pricing: VenuePricing[];
    rating: number;
    reviewCount: number;
    photos: string[];
    isVerified: boolean;
    createdAt: string;
    updatedAt: string;
    stats: VenueStats;
    features: {
        hasParking: boolean;
        isWheelchairAccessible: boolean;
        hasFood: boolean;
        hasBar: boolean;
        hasTournaments: boolean;
        hasTraining: boolean;
    };
    rules: {
        minimumAge?: number;
        dresscode?: string;
        reservationRequired: boolean;
        membershipRequired: boolean;
    };
}

export interface VenueEvent {
    id: string;
    venueId: string;
    name: string;
    description: string;
    eventType: 'tournament' | 'league' | 'social' | 'training';
    startTime: string;
    endTime: string;
    registrationDeadline?: string;
    maxParticipants?: number;
    currentParticipants: number;
    entryFee?: number;
    prizePool?: number;
    status: 'upcoming' | 'in_progress' | 'completed' | 'cancelled';
    participants: {
        id: string;
        name: string;
        registeredAt: string;
    }[];
    createdAt: string;
    updatedAt: string;
}

export interface LeaderboardEntry {
    id: string;
    userId: string;
    name: string;
    rank: number;
    points: number;
    gamesPlayed: number;
    gamesWon: number;
    winRate: number;
    averageScore: number;
    highestScore: number;
    streak: number;
    lastPlayed: string;
} 

File: src/dojopool/schemas/venue.py
"""Schemas for venue-related data validation."""

from marshmallow import Schema, ValidationError, fields, validate, validates


class VenueSchema(Schema):
    """Schema for venue data."""

    id = fields.Int(dump_only=True)
    name = fields.Str(required=True, validate=validate.Length(min=1, max=100))
    description = fields.Str(validate=validate.Length(max=500))
    address = fields.Str(required=True, validate=validate.Length(min=1, max=200))
    city = fields.Str(required=True, validate=validate.Length(min=1, max=100))
    state = fields.Str(required=True, validate=validate.Length(min=1, max=100))
    country = fields.Str(required=True, validate=validate.Length(min=1, max=100))
    postal_code = fields.Str(required=True, validate=validate.Length(min=1, max=20))
    latitude = fields.Float()
    longitude = fields.Float()
    phone = fields.Str(validate=validate.Length(max=20))
    email = fields.Email(validate=validate.Length(max=100))
    website = fields.URL(validate=validate.Length(max=200))
    hours = fields.Dict(keys=fields.Str(), values=fields.Dict())
    features = fields.List(fields.Str())
    tables = fields.Int(required=True, validate=validate.Range(min=1))
    pricing = fields.Dict(keys=fields.Str(), values=fields.Float())
    rating = fields.Float(dump_only=True)
    total_ratings = fields.Int(dump_only=True)
    is_verified = fields.Bool(dump_only=True)
    is_active = fields.Bool(dump_only=True)
    created_at = fields.DateTime(dump_only=True)
    updated_at = fields.DateTime(dump_only=True)

    @validates("latitude")
    def validate_latitude(self, value):
        """Validate latitude is within valid range."""
        if value and (value < -90 or value > 90):
            raise ValidationError("Latitude must be between -90 and 90")

    @validates("longitude")
    def validate_longitude(self, value):
        """Validate longitude is within valid range."""
        if value and (value < -180 or value > 180):
            raise ValidationError("Longitude must be between -180 and 180")


class VenueCheckInSchema(Schema):
    """Schema for venue check-in data."""

    id = fields.Int(dump_only=True)
    venue_id = fields.Int(dump_only=True)
    user_id = fields.Int(dump_only=True)
    table_number = fields.Int(validate=validate.Range(min=1))
    game_type = fields.Str(validate=validate.Length(max=50))
    checked_in_at = fields.DateTime(dump_only=True)
    checked_out_at = fields.DateTime(dump_only=True)


class VenueLeaderboardSchema(Schema):
    """Schema for venue leaderboard data."""

    id = fields.Int(dump_only=True)
    venue_id = fields.Int(dump_only=True)
    user_id = fields.Int(dump_only=True)
    points = fields.Int(dump_only=True)
    wins = fields.Int(dump_only=True)
    losses = fields.Int(dump_only=True)
    current_streak = fields.Int(dump_only=True)
    highest_streak = fields.Int(dump_only=True)
    last_played = fields.DateTime(dump_only=True)


class VenueEventSchema(Schema):
    """Schema for venue event data."""

    id = fields.Int(dump_only=True)
    venue_id = fields.Int(dump_only=True)
    name = fields.Str(required=True, validate=validate.Length(min=1, max=100))
    description = fields.Str(validate=validate.Length(max=500))
    event_type = fields.Str(required=True, validate=validate.Length(min=1, max=50))
    start_time = fields.DateTime(required=True)
    end_time = fields.DateTime(required=True)
    registration_deadline = fields.DateTime()
    max_participants = fields.Int(validate=validate.Range(min=1))
    entry_fee = fields.Float(validate=validate.Range(min=0))
    prize_pool = fields.Float(validate=validate.Range(min=0))
    status = fields.Str(dump_only=True)
    created_at = fields.DateTime(dump_only=True)
    updated_at = fields.DateTime(dump_only=True)

    @validates("end_time")
    def validate_end_time(self, value):
        """Validate end time is after start time."""
        if (
            value
            and self.context.get("start_time")
            and value <= self.context["start_time"]
        ):
            raise ValidationError("End time must be after start time")

    @validates("registration_deadline")
    def validate_registration_deadline(self, value):
        """Validate registration deadline is before start time."""
        if (
            value
            and self.context.get("start_time")
            and value >= self.context["start_time"]
        ):
            raise ValidationError("Registration deadline must be before start time")
